Testing: Refactor CI / Test Infrastructure to use Runtime Mounted Source Code and Registry Caching in Autotests.

0xquark opened the pull request:
<!-- Please read https://rucio.cern.ch/documentation/contributing before submitting a pull request -->
Ref: #7839 #7840 

This PR updates the CI testing approach by switching from building Docker images with the source baked in to using pre-built base runtime images and mounting the PR’s source code at runtime in `autotests.yml`.

- The base images (for Python 3.9 and 3.10) are built and pushed to GHCR using `runtime.Dockerfile` and the `runtime_images.yml` workflow (only if the image isn’t already available).

- These runtime images are then used by autotests.yml to execute tests. Inside `run_tests.py`, `test.sh`, and `docker-compose.yml`, the paths are configured to use the mounted PR code, which is installed in editable mode via `pip install -e`.

- Image builds uses registry based caching (rebuilding only when dependencies change), otherwise pulling the existing image. In case of dependencies bump or change, It uses old layers from the cache which hasn't changed.

The whole workflow with caching takes around [30-35 mins](https://github.com/0xquark/rucio/actions/runs/16190100219) instead of around [1h](https://github.com/rucio/rucio/actions/runs/16216539701). This happens because we only build the image once so we save a lot of time while running other tests in the matrix using the same image.

The whole testing workflow for autotests is described below:
<img width="2278" height="973" alt="image" src="https://github.com/user-attachments/assets/28a293f1-5f3a-46b8-aed9-68cd6427c918" />

0xquark committed:
testing: Refactor CI to mount rucio source code for testing; ref #7839,#7849

This PR updates the CI testing approach by switching from building Docker images with the source baked in to using pre-built base runtime images and mounting the PR’s source code at runtime in `autotests.yml`.

- The base images (for Python 3.9 and 3.10) are built and pushed to GHCR using `runtime.Dockerfile`and the `runtime_images.yml` workflow (only if the image isn’t already available).

- These runtime images are then used by autotests.yml to execute tests. Inside `run_tests.py`, `test.sh`, and `docker-compose.yml`, the paths are configured to use the mounted PR code, which is installed in editable mode via `pip install -e`.

- Image builds uses registry based caching (rebuilding only when dependencies change), otherwise pulling the existing image.

0xquark committed:
testing: Add cleanup workflow to periodically delete stale runtime images

0xquark committed:
fix: Use a new build strategy in order to avoid permission issue with
packages

0xquark commented:
Updated Workflow (with fix to the permission issue related to packages):
<img width="1898" height="928" alt="image" src="https://github.com/user-attachments/assets/c7da56e5-f0a8-46ca-b9d4-a24b55591977" />

0xquark commented:
This workflow has been tested in [this PR](https://github.com/rucio/rucio/pull/7870) on @rdimaio’s fork.

- When the workflow runs on a `push`, since it doesn’t find the image on GHCR and the PR targets the master branch. In this case, the `build_locally` flag is set to `true` (as we would not have write permission for the actions run on main repository), so the Autotests run using a locally built image within the `rucio/rucio` GitHub Actions. 

- On Riccardo’s fork, Actions don’t run for pull requests targeting `master`. Since it’s the first time the image is being built there, the workflow builds and pushes the image to the fork’s package registry (as the fork has write permissions). Here, the`build_locally` flag remains `false` because it’s not running on the PR targeted to `rucio/rucio master` branch, and the GHCR image isn’t yet available. [These Actions on the fork build the image once and reuse it across jobs](https://github.com/rdimaio/rucio/actions/runs/16368355030/job/46250741137).

- To confirm that the PR code is properly mounted, we intentionally introduced incorrect assertions to break tests. This successfully triggered failures, as seen in [this commit](https://github.com/rucio/rucio/pull/7870/commits/36ccac4e2b5fa55d4e6d4a973f4c7f53ffff2489). The failures occurred in both the [main Rucio Actions run](https://github.com/rucio/rucio/actions/runs/16369793381/job/46255249844?pr=7870) and the [fork’s Actions run](https://github.com/rdimaio/rucio/actions/runs/16369792548/job/46255254182). On the fork, the workflow [pulls the pushed image from GHCR](https://github.com/rdimaio/rucio/actions/runs/16369792548/job/46255238903) instead of rebuilding it. 

These tests cover most of the key scenarios for the workflow and confirm that everything works as intended. There shouldn’t be any issues after merging the PR. Thanks @rdimaio for helping me test on his fork.

rdimaio commented on .github/workflows/autotest.yml:
```diff
@@ -160,31 +167,56 @@ jobs:
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+        continue-on-error: true
+      - name: Build Python 3.9 image locally
+        if: needs.runtime_images.outputs.build_locally == 'true' && matrix.cfg.PYTHON == '3.9'
+        uses: docker/build-push-action@263435318d21b8e681c14492fe198d362a7d2c83 # v6.18.0
+        with:
+          context: .
+          file: etc/docker/test/runtime.Dockerfile
+          target: final
+          build-args: |
+            PYTHON=3.9
+          push: false
+          tags: ${{ needs.runtime_images.outputs.py39_image }}
+          outputs: type=docker
+          cache-from: type=registry,ref=ghcr.io/${{ github.repository }}/rucio-dev-runtime:py39-buildcache
+        continue-on-error: false
+      - name: Build Python 3.10 image locally
+        if: needs.runtime_images.outputs.build_locally == 'true' && matrix.cfg.PYTHON == '3.10'
+        uses: docker/build-push-action@263435318d21b8e681c14492fe198d362a7d2c83 # v6.18.0
+        with:
+          context: .
+          file: etc/docker/test/runtime.Dockerfile
+          target: final
+          build-args: |
+            PYTHON=3.10
+          push: false
+          tags: ${{ needs.runtime_images.outputs.py310_image }}
+          outputs: type=docker
+          cache-from: type=registry,ref=ghcr.io/${{ github.repository }}/rucio-dev-runtime:py310-buildcache
+        continue-on-error: false
```
Could this be simplified in some way? e.g. since we have `matrix.cfg.PYTHON` that provides the version number, can we pass that directly to the `build-args` `PYTHON` argument and use it in the build cache names (e.g. `rucio-dev-runtime:py{{ matrix.cfg.PYTHON }}-buildcache` so that we can simplify this in a single step? 

I know that ultimately we only run one of these two steps depending on whether we're building python 3.9 or 3.10, but these two steps seem to mostly be the same, so I'm wondering if they can be simplified like that

maany replied:
@rdimaio that will be part of autotest simplification :) https://github.com/maany/rucio/blob/861bf9a7379cfe59754be940674aeacc7805deed/.github/workflows/simple-autotest.yml

rdimaio commented on etc/docker/test/runtime.Dockerfile:
```diff
@@ -0,0 +1,181 @@
+    ENV PYTHON_310_PATCH_VERSION="4"
+    ENV RUCIO_HOME="/opt/rucio"
+
+FROM base AS oracle-client
+    RUN dnf install -y libnsl libaio nodejs npm
+    RUN rpm -i https://download.oracle.com/otn_software/linux/instantclient/1912000/oracle-instantclient19.12-basiclite-19.12.0.0.0-1.x86_64.rpm && \
+        echo "/usr/lib/oracle/19.12/client64/lib" > /etc/ld.so.conf.d/oracle-instantclient.conf;
```
No longer needed https://github.com/rucio/containers/pull/421 - @Geogouz pinging you just to confirm

rdimaio replied:
Checked with @Geogouz - let's leave it in for now

rdimaio commented on .github/workflows/runtime_images.yml:
```diff
@@ -0,0 +1,181 @@
+            rucio.image.build.run_id=${{ github.run_id }}
+          cache-from: type=registry,ref=ghcr.io/${{ github.repository }}/rucio-dev-runtime:py310-buildcache
+          cache-to: type=registry,ref=ghcr.io/${{ github.repository }}/rucio-dev-runtime:py310-buildcache,mode=max
+        continue-on-error: true
```
can this be simplified in a way to avoid duplication?

0xquark replied:
Let me try this by using a [Matrix](https://docs.github.com/en/actions/how-tos/write-workflows/choose-what-workflows-do/run-job-variations) for the versions, this way i think this could be more simplified. Would probably reduce a lot of boilerplate.

0xquark replied:
I tried using the Matrix method to simplify things by setting it up for both Versions 3.9 and 3.10. The main advantage of using Matrix is that jobs run at parallel, so we could save about 3-4 minutes because both images would start building together at same time.

But after trying it, I found out that Matrix jobs run separately and independent of each other so each job can overwrite the other’s output depending on which finishes last. This causes problems because only one image output will be saved, and the other will be empty. That makes it hard to pass the built image (with its hash) to the Autotests and other tests.

I looked for a solution, but it didn’t work because this feature [isn’t supported yet in GitHub runners and is requested](https://github.com/orgs/community/discussions/17245). Another idea was to save the image tag as an artifact and then download it in another job to combine the outputs. But this makes the setup more complicated and not worth the extra complexity.

0xquark replied:
The normal way to use loop for both the builds (3.9, 3.10) works well for simplification. It reduced a lot of code and is now much more cleaner and is part of a single step:).
https://github.com/0xquark/rucio/actions/runs/16569563583/job/46857899016

0xquark committed:
refactor: Simplify runtime image build workflow

Couldn't use matrix for simplification (see https://github.com/rucio/rucio/pull/7853#discussion_r2236221956)
I've tried to simplify by doing a loop for 3.9, 3.10 in a single build step rather then separate steps.

rdimaio reviewed (commented):
Please rebase to fix merge conflicts

0xquark committed:
testing: bump checkout gh action version v4.1.7-> v5.0.0

0xquark committed:
testing: Fix argument handling using relative paths and filter  BytesWarning in oracle autotests.

After rebasing from the master, The CI was failing due to missing fixtures (found in conftest.py), It seems that due to this commit https://github.com/rucio/rucio/commit/5035e4d04dae5d390ab1a58e0a479ccd97eadac2 , The conftest was loaded from the mounted source in my case i.e `/rucio_source/tests/conftest.py`  but the test ran from `/opt/rucio/tests/ ` instead of `/rucio_source/tests/`. The issue is pytest loaded conftest.py from one location but tried to run tests from another location which caused the CI to fail due to missing fixtures.
I introduced the fix by changing the test paths from absolute to relative path therefore it would not depend on what source that you mount.

- This commit also adds     ignore::BytesWarning to fix autotests failing due to this thick mode warning introduced in https://github.com/rucio/rucio/commit/c4df1306d192b518b38e023e783ea998c6c0dfb5
